File: src/familiar_connect/processors/fact_supersede_worker.py
"""Watermark-driven fact supersession worker.

For each newly-appended fact, asks background LLM whether it
retires any prior current facts about the same subject — if so,
calls :meth:`HistoryStore.supersede` (existing-id form) to repoint
them at the new fact.

System-time bookkeeping half of the fact lifecycle. ``valid_to``
(world-time) is left to extractor and to speaker who anchors a
real-world end. Supersession runs slower than extractor (60 s
default) — fact churn bounded by conversation pace, missing a tick
of LLM-assisted retirement is harmless.
"""

from __future__ import annotations

import asyncio
import logging
from typing import TYPE_CHECKING

from familiar_connect import log_style as ls
from familiar_connect.diagnostics.spans import span
from familiar_connect.llm import Message
from familiar_connect.structured_output import coerce_positive_int_list
from familiar_connect.structured_request import (
    Field,
    Schema,
    render_contract,
    request_structured,
)

if TYPE_CHECKING:
    from typing import Any

    from familiar_connect.history.async_store import AsyncHistoryStore
    from familiar_connect.history.store import Fact
    from familiar_connect.llm import LLMClient

_logger = logging.getLogger("familiar_connect.processors.fact_supersede_worker")

# Reply-shape contract for "which priors does this new fact retire" —
# declared, rendered + parsed via :mod:`familiar_connect.structured_request`
# (#167) instead of a hand-typed JSON string in the prompt builder.
_SUPERSEDE_SCHEMA = Schema(
    fields=(Field("superseded_ids", "[<id>...]"),),
    root="object",
    empty_note="Empty list when nothing is retired.",
    constraints=(
        "Only include ids from the Prior facts list below — do not invent ids.",
    ),
)


class FactSupersedeWorker:
    """Retires prior facts replaced by newer ones about the same subject.

    :param batch_size: max new facts per tick. Each gets one LLM
        call, batched per subject.
    :param tick_interval_s: idle interval. 60 s default — slower
        than 15 s extractor since supersession isn't latency-critical
        and adds an LLM call per new fact.
    :param priors_max: cap on prior facts shown to LLM per subject.
        Prevents prompt bloat for long-running familiars with many
        priors per person.
    """

    name: str = "fact-supersede-worker"

    def __init__(
        self,
        *,
        store: AsyncHistoryStore,
        llm_client: LLMClient,
        familiar_id: str,
        batch_size: int = 5,
        tick_interval_s: float = 60.0,
        priors_max: int = 20,
    ) -> None:
        self._store = store
        self._llm = llm_client
        self._familiar_id = familiar_id
        self._batch_size = max(1, batch_size)
        self._tick_interval_s = tick_interval_s
        self._priors_max = max(1, priors_max)
        self._last_seen_fact_id: int = 0

    def prime_watermark(self) -> None:
        """Skip historical backlog; start from latest current id.

        Called by projector factory at process start so fresh deploy
        doesn't burn LLM calls re-evaluating every old fact on first
        tick. Tests may call directly to assert no-new-facts path.
        """
        self._last_seen_fact_id = self._store.sync.latest_fact_id(
            familiar_id=self._familiar_id
        )

    async def run(self) -> None:
        """Forever loop; tick on interval. Cancel to stop."""
        self.prime_watermark()
        while True:
            try:
                await self.tick()
            except asyncio.CancelledError:
                raise
            except Exception as exc:  # noqa: BLE001 — worker must not die
                _logger.warning(
                    f"{ls.tag('Supersede', ls.R)} "
                    f"{ls.kv('tick_error', repr(exc), vc=ls.R)}"
                )
            await asyncio.sleep(self._tick_interval_s)

    @span("fact_supersede.tick")
    async def tick(self) -> int:
        """Evaluate up to ``batch_size`` new current facts; return supersedes."""
        candidates = await self._store.recent_facts(
            familiar_id=self._familiar_id,
            limit=self._batch_size,
            include_superseded=False,
        )
        new = [f for f in candidates if f.id > self._last_seen_fact_id]
        if not new:
            return 0

        # Oldest-new first so cascading retirements settle deterministically
        new.sort(key=lambda f: f.id)
        retired = 0
        for f_new in new:
            retired += await self._evaluate(f_new)

        # Advance watermark to highest id seen this tick (even on bad
        # llm output) — prevents loops on a fact whose prompt model
        # consistently fails to parse
        if candidates:
            self._last_seen_fact_id = max(f.id for f in candidates)

        if retired:
            _logger.info(
                f"{ls.tag('Supersede', ls.LM)} "
                f"{ls.kv('evaluated', str(len(new)), vc=ls.LW)} "
                f"{ls.kv('retired', str(retired), vc=ls.LM)} "
                f"{ls.kv('watermark', str(self._last_seen_fact_id), vc=ls.LW)}"
            )
        return retired

    async def _evaluate(self, f_new: Fact) -> int:
        """Ask LLM which priors `f_new` retires; supersede them; return count."""
        if not f_new.subjects:
            return 0
        retired = 0
        seen_priors: set[int] = set()
        for subject in f_new.subjects:
            priors = await self._store.facts_for_subject(
                familiar_id=self._familiar_id,
                canonical_key=subject.canonical_key,
                include_superseded=False,
            )
            # Exclude f_new itself, dedupe across subjects, cap to priors_max
            unique = [
                p for p in priors if p.id != f_new.id and p.id not in seen_priors
            ][-self._priors_max :]
            if not unique:
                continue
            seen_priors.update(p.id for p in unique)

            prompt = _build_supersede_prompt(f_new=f_new, priors=unique)
            result = await request_structured(
                self._llm, messages=prompt, schema=_SUPERSEDE_SCHEMA
            )
            ids = _superseded_ids(result.value, valid={p.id for p in unique})
            if not ids:
                continue
            # Existing-id form: repoint each old row at f_new (mints
            # nothing). Per-id skip-and-record — a prior already retired
            # by an earlier subject lands in `skipped`, not an exception.
            result = await self._store.supersede(
                familiar_id=self._familiar_id,
                obsolete_facts=ids,
                new_fact=f_new,
            )
            retired += len(result.superseded)
        return retired


def _build_supersede_prompt(*, f_new: Fact, priors: list[Fact]) -> list[Message]:
    """LLM prompt: which priors does ``f_new`` replace."""
    persona = (
        "You decide whether a new fact retires earlier facts about the "
        "same person. A fact is *retired* when the new one contradicts "
        "or directly replaces it (e.g., 'Alice loves hiking' is retired "
        'by "Alice now hates hiking"). A fact is NOT retired just '
        "because it's older or differently worded — facts about "
        "independent topics coexist."
    )
    header = f"{persona}\n\n{render_contract(_SUPERSEDE_SCHEMA)}"
    lines: list[str] = [
        f"New fact (id={f_new.id}): {f_new.text}",
        "",
        "Prior facts:",
    ]
    lines.extend(f"- id={p.id}: {p.text}" for p in priors)
    return [
        Message(role="system", content=header),
        Message(role="user", content="\n".join(lines)),
    ]


def _superseded_ids(value: Any, *, valid: set[int]) -> list[int]:  # noqa: ANN401 — parsed JSON
    """Distinct prior ids the model marked superseded, filtered to *valid*.

    *value* is the parsed reply object from
    :func:`familiar_connect.structured_request.request_structured`; a
    non-object, missing key, or non-list value all degrade to ``[]``.
    """
    if not isinstance(value, dict):
        return []
    return [
        i for i in coerce_positive_int_list(value.get("superseded_ids")) if i in valid
    ]

File: familiar-connect/src/processors/fact_supersede_worker.rs
//! FactSupersedeWorker retirement (subsystem 07; Python `processors/fact_supersede_worker.py`).
//!
//! For each newly-appended fact, ask the background LLM whether it retires any
//! prior current facts about the same subject — if so, call
//! [`AsyncHistoryStore::supersede`] (existing-id form) to repoint them at the
//! new fact. This is the *system-time* half of the fact lifecycle; `valid_to`
//! (world-time) is left to the extractor and to a speaker who anchors a
//! real-world end.
//!
//! The watermark (`last_seen_fact_id`) is **in-memory only** — it starts at 0
//! and is primed to `latest_fact_id` at run start so a fresh deploy never burns
//! LLM calls re-evaluating historical facts. State is lost on restart and
//! re-primed to "latest", so facts appended while the process was down are never
//! supersede-evaluated (accepted: this bookkeeping is best-effort).

use std::collections::HashSet;
use std::sync::Arc;
use std::sync::LazyLock;
use std::sync::atomic::{AtomicI64, Ordering};
use std::time::Duration;

use serde_json::Value;
use tokio_util::sync::CancellationToken;

use crate::diagnostics::spans::timed_async;
use crate::history::async_store::AsyncHistoryStore;
use crate::history::store::{Fact, NewFact};
use crate::llm::{LlmClient, Message};
use crate::log_style as ls;
use crate::structured_output::coerce_positive_int_list;
use crate::structured_request::{
    DEFAULT_MAX_RETRIES, Field, Schema, render_contract, request_structured,
};

/// Log/task label + registry name for this projector.
const NAME: &str = "fact-supersede-worker";

/// Reply-shape contract for "which priors does this new fact retire".
static SUPERSEDE_SCHEMA: LazyLock<Schema> = LazyLock::new(|| {
    Schema::object(vec![Field::new("superseded_ids", "[<id>...]")])
        .with_empty_note("Empty list when nothing is retired.")
        .with_constraints(vec![
            "Only include ids from the Prior facts list below — do not invent ids.".to_string(),
        ])
});

/// Retires prior facts replaced by newer ones about the same subject.
pub struct FactSupersedeWorker {
    store: Arc<AsyncHistoryStore>,
    llm: Arc<dyn LlmClient>,
    familiar_id: String,
    batch_size: i64,
    priors_max: usize,
    tick_interval: Duration,
    last_seen_fact_id: AtomicI64,
}

impl FactSupersedeWorker {
    /// Construct with the required handles; knobs default per spec
    /// (`batch_size = 5`, `tick_interval_s = 60.0`, `priors_max = 20`).
    #[must_use]
    pub fn new(
        store: Arc<AsyncHistoryStore>,
        llm: Arc<dyn LlmClient>,
        familiar_id: impl Into<String>,
    ) -> Self {
        Self {
            store,
            llm,
            familiar_id: familiar_id.into(),
            batch_size: 5,
            priors_max: 20,
            tick_interval: Duration::from_secs_f64(60.0),
            last_seen_fact_id: AtomicI64::new(0),
        }
    }

    /// Max new facts per tick (clamped to `>= 1`).
    #[must_use]
    pub const fn batch_size(mut self, batch_size: i64) -> Self {
        self.batch_size = if batch_size < 1 { 1 } else { batch_size };
        self
    }

    /// Cap on prior facts shown to the LLM per subject (clamped to `>= 1`).
    #[must_use]
    pub fn priors_max(mut self, priors_max: i64) -> Self {
        self.priors_max = usize::try_from(priors_max.max(1)).unwrap_or(usize::MAX);
        self
    }

    /// Idle-loop interval in seconds.
    #[must_use]
    pub fn tick_interval_s(mut self, secs: f64) -> Self {
        self.tick_interval = Duration::from_secs_f64(secs);
        self
    }

    /// The projector's log/task label.
    #[must_use]
    pub const fn name(&self) -> &'static str {
        NAME
    }

    /// Skip the historical backlog; start from the latest current fact id.
    ///
    /// Called by the projector factory at process start (and directly by tests
    /// to assert the no-new-facts path) so a fresh deploy doesn't burn LLM calls
    /// re-evaluating every old fact on the first tick. SYNC — reads
    /// `store.sync().latest_fact_id`, mirroring the Python `.sync` escape hatch.
    ///
    /// # Errors
    /// Propagates a [`crate::history::StoreError`] if the underlying read fails.
    pub fn prime_watermark(&self) -> Result<(), crate::history::StoreError> {
        let latest = self.store.sync().latest_fact_id(&self.familiar_id)?;
        self.last_seen_fact_id.store(latest, Ordering::SeqCst);
        Ok(())
    }

    /// Forever loop; primes the watermark, then ticks on interval. Cancel to stop.
    pub async fn run(&self, cancel: CancellationToken) {
        if let Err(exc) = self.prime_watermark() {
            tracing::warn!(
                target: "familiar_connect.processors.fact_supersede_worker",
                "{} {}",
                ls::tag("Supersede", ls::R),
                ls::kv_styled("prime_error", &format!("{exc:?}"), ls::W, ls::R),
            );
        }
        loop {
            if cancel.is_cancelled() {
                break;
            }
            if let Err(exc) = self.tick().await {
                tracing::warn!(
                    target: "familiar_connect.processors.fact_supersede_worker",
                    "{} {}",
                    ls::tag("Supersede", ls::R),
                    ls::kv_styled("tick_error", &format!("{exc:?}"), ls::W, ls::R),
                );
            }
            tokio::select! {
                () = cancel.cancelled() => break,
                () = tokio::time::sleep(self.tick_interval) => {}
            }
        }
    }

    /// Evaluate up to `batch_size` new current facts; return the count retired.
    pub async fn tick(&self) -> anyhow::Result<i64> {
        timed_async("fact_supersede.tick", async move {
            let candidates = self
                .store
                .recent_facts(self.familiar_id.clone(), self.batch_size, false, None)
                .await?;
            let watermark = self.last_seen_fact_id.load(Ordering::SeqCst);
            let mut new: Vec<Fact> = candidates
                .iter()
                .filter(|f| f.id > watermark)
                .cloned()
                .collect();
            if new.is_empty() {
                return Ok(0);
            }
            // Oldest-new first so cascading retirements settle deterministically.
            new.sort_by_key(|f| f.id);
            let mut retired: i64 = 0;
            for f_new in &new {
                retired += self.evaluate(f_new).await?;
            }
            // Advance the watermark to the highest id seen this tick — even on
            // garbage LLM output — so a consistently-unparseable fact can't loop.
            if let Some(max_id) = candidates.iter().map(|f| f.id).max() {
                self.last_seen_fact_id.store(max_id, Ordering::SeqCst);
            }
            if retired > 0 {
                tracing::info!(
                    target: "familiar_connect.processors.fact_supersede_worker",
                    "{} {} {} {}",
                    ls::tag("Supersede", ls::LM),
                    ls::kv_styled("evaluated", &new.len().to_string(), ls::W, ls::LW),
                    ls::kv_styled("retired", &retired.to_string(), ls::W, ls::LM),
                    ls::kv_styled(
                        "watermark",
                        &self.last_seen_fact_id.load(Ordering::SeqCst).to_string(),
                        ls::W,
                        ls::LW,
                    ),
                );
            }
            Ok(retired)
        })
        .await
    }

    /// Ask the LLM which priors `f_new` retires; supersede them; return the count.
    async fn evaluate(&self, f_new: &Fact) -> anyhow::Result<i64> {
        if f_new.subjects.is_empty() {
            return Ok(0);
        }
        let mut retired: i64 = 0;
        let mut seen_priors: HashSet<i64> = HashSet::new();
        for subject in &f_new.subjects {
            let priors = self
                .store
                .facts_for_subject(
                    self.familiar_id.clone(),
                    subject.canonical_key.clone(),
                    0,
                    false,
                    None,
                )
                .await?;
            // Exclude f_new itself, dedupe across subjects, cap to the last
            // `priors_max` (most recent).
            let filtered: Vec<&Fact> = priors
                .iter()
                .filter(|p| p.id != f_new.id && !seen_priors.contains(&p.id))
                .collect();
            let start = filtered.len().saturating_sub(self.priors_max);
            let unique = &filtered[start..];
            if unique.is_empty() {
                continue;
            }
            for p in unique {
                seen_priors.insert(p.id);
            }
            let valid: HashSet<i64> = unique.iter().map(|p| p.id).collect();

            let prompt = build_supersede_prompt(f_new, unique);
            let result = request_structured(
                self.llm.as_ref(),
                &prompt,
                &SUPERSEDE_SCHEMA,
                DEFAULT_MAX_RETRIES,
            )
            .await?;
            let ids = superseded_ids(result.value.as_ref(), &valid);
            if ids.is_empty() {
                continue;
            }
            // Existing-id form: repoint each old row at f_new (mints nothing).
            // Per-id skip-and-record — a prior already retired by an earlier
            // subject lands in `skipped`, not an exception.
            let outcome = self
                .store
                .supersede(self.familiar_id.clone(), ids, NewFact::Repoint(f_new.id))
                .await?;
            retired += i64::try_from(outcome.superseded.len()).unwrap_or(i64::MAX);
        }
        Ok(retired)
    }
}

/// LLM prompt: which priors does `f_new` replace?
fn build_supersede_prompt(f_new: &Fact, priors: &[&Fact]) -> Vec<Message> {
    let persona = "You decide whether a new fact retires earlier facts about the \
        same person. A fact is *retired* when the new one contradicts \
        or directly replaces it (e.g., 'Alice loves hiking' is retired \
        by \"Alice now hates hiking\"). A fact is NOT retired just \
        because it's older or differently worded — facts about \
        independent topics coexist.";
    let header = format!("{persona}\n\n{}", render_contract(&SUPERSEDE_SCHEMA));
    let mut lines: Vec<String> = vec![
        format!("New fact (id={}): {}", f_new.id, f_new.text),
        String::new(),
        "Prior facts:".to_string(),
    ];
    lines.extend(priors.iter().map(|p| format!("- id={}: {}", p.id, p.text)));
    vec![
        Message::new("system", header),
        Message::new("user", lines.join("\n")),
    ]
}

/// Distinct prior ids the model marked superseded, filtered to `valid`.
///
/// A non-object, missing key, or non-list value all degrade to `[]`.
fn superseded_ids(value: Option<&Value>, valid: &HashSet<i64>) -> Vec<i64> {
    let Some(Value::Object(map)) = value else {
        return Vec::new();
    };
    let raw = map.get("superseded_ids").cloned().unwrap_or(Value::Null);
    coerce_positive_int_list(&raw)
        .into_iter()
        .filter(|i| valid.contains(i))
        .collect()
}

#[cfg(test)]
mod tests {
    use super::superseded_ids;
    use serde_json::json;
    use std::collections::HashSet;

    fn valid(ids: &[i64]) -> HashSet<i64> {
        ids.iter().copied().collect()
    }

    #[test]
    fn non_object_degrades_to_empty() {
        assert!(superseded_ids(Some(&json!([1, 2])), &valid(&[1, 2])).is_empty());
        assert!(superseded_ids(Some(&json!("x")), &valid(&[1])).is_empty());
        assert!(superseded_ids(None, &valid(&[1])).is_empty());
    }

    #[test]
    fn missing_key_degrades_to_empty() {
        assert!(superseded_ids(Some(&json!({"other": [1]})), &valid(&[1])).is_empty());
    }

    #[test]
    fn filters_to_valid_and_drops_hallucinated() {
        // 9999 is not in the shown prior set; the new fact's own id would be
        // filtered the same way.
        assert_eq!(
            superseded_ids(
                Some(&json!({"superseded_ids": [3, 9999, 4]})),
                &valid(&[3, 4])
            ),
            vec![3, 4]
        );
    }

    #[test]
    fn coercion_rejects_bools_and_dedups() {
        // JSON booleans never coerce to 1; duplicates collapse.
        assert_eq!(
            superseded_ids(
                Some(&json!({"superseded_ids": [true, 3, "3", "4", 4]})),
                &valid(&[3, 4])
            ),
            vec![3, 4]
        );
    }
}
